### PR TITLE
[MRG+1] raw.annotations should be emtpy list not None

### PR DESCRIPTION
**Pull request by massich**

### CODE REVIEW

**Comment by mmagnuski** [comment]
@massich - Could you paste your concerns here? They will get lost in the comment in the other PR.

**Comment by massich** [comment]
sure.
It is true that `xx.annotations is None` comes really natural. But then imagine the following scenario:

```py
@testing.requires_testing_data
def test_xx():

    #################################
    # Adapt the reference raw object
    #################################

    raw = read_raw_fif(fif_fname, verbose=0)
    raw_data_offset_in_sec = raw.first_samp / raw.info['sfreq']
    data_start_from_origin_in_sec = 50
    data_duration_in_sec = 10
    crop_tmin = data_start_from_origin_in_sec - raw_data_offset_in_sec
    crop_tmax = crop_tmin + data_duration_in_sec

    reference_mne_data = raw.copy().crop(tmin=crop_tmin, tmax=crop_tmax)

    # XXX There are more tests to be done to ensure that the reference object
    # is proper
    assert reference_mne_data.first_samp > raw.first_samp
    # Ensure first_samp
    expected_first_samp = data_start_from_origin_in_sec * raw.info['sfreq']
    assert reference_mne_data.first_samp == ceil(expected_first_samp) # XXX This bugs me

    assert reference_mne_data.annotations is None
    # Add some annotations
    orig_time = reference_mne_data.info['meas_date']
    reference_annotations = Annotations(onset=[51.2],
                                        duration=[1.0],
                                        description=['a'],
                                        orig_time=orig_time)
    reference_mne_data.set_annotations(reference_annotations)

    #################################
    # Split the data into 3 objects
    #################################

    split_duration_in_sec = data_duration_in_sec / 3
    part_a = reference_mne_data.copy().crop(tmin=0, tmax=split_duration_in_sec)
    part_b = reference_mne_data.copy().crop(tmin=split_duration_in_sec,
                                            tmax=2 * split_duration_in_sec)
    part_c = reference_mne_data.copy().crop(tmin=2 * split_duration_in_sec,
                                            tmax=None)

    # This seems a natural way to be called
    #
    assert part_a.annotations is not None
    assert part_b.annotations is None
    assert part_c.annotations is None


    # But I might also like to do that
    #
    n_annotations = [p.annotations.length() for p in (part_a, part_b, part_c)]
    assert n_annotations == [1, 0, 0]
```

you either have one or the other. But you cannot have both, because the `is` operator cannot be overwritten. 

The problem is that once you assign to `None` you are no longer an `Annotation`. 
The concept of `None` is not an empty `Annotation` but a nothing object.

and I don't seem to be able to figure out which is the nicest of the two approaches, from the point of view of the user.

**Comment by massich** [comment]
The thing is that I've not made up my mind yet. And therefore I've let it as it is, work on the readers and if I find an elegant solution I'll come back to this.

Of course, I'm completely open to any suggestions.

**Comment by mmagnuski** [comment]
From my point of view it is equivalently easy to check `annot is not None`, `annot.length() > 1` or `len(annot.onsets)` but having empty annotations instead of `None` is much more convenient when looping through files. I'm not sure if this is common enough scenario to worry, but I am starting to think it would be clearer to always have an Annotation object in `.annotations` attribute, and have it empty if there are no annotations.

**Comment by larsoner** [comment]
Currently IIRC if `raw.annot is None` we don't write anything to disk, otherwise we do even if `len(raw.annot) == 0`, because:

- the `raw.annot` stores both events *and* an `orig_time`, so even if `len(raw.annot) == 0` there can still be some information left that a user supplied. And this should persist even on I/O because our check is just whether or not it's `None` not its length:

  https://github.com/mne-tools/mne-python/blob/master/mne/io/base.py#L2355

- you can also tell if a raw object had `Annotations` added or not by whether or not it is `None`.

- state is preserved nicely and simply on round-trip I/O

Thus if an object has annotations (`raw.annot is not None`) after any cropping operation it should still be non-None, but can have `len(raw.annot) == 0`. So I would say these last two `assert` statements are wrong:
```
    assert part_a.annotations is not None
    assert part_b.annotations is None
    assert part_c.annotations is None
```
This should be the case in `master` and it should be preserved. If it's not the case in `master`, I think we have a bug/inconsistency and it should be fixed.

**Comment by larsoner** [comment]
... in other words, I'm -1 on this PR

**Comment by agramfort** [comment]
@larsoner what triggered this PR is this:

>
mne/io/egi/egimff.py:
self.set_annotations(Annotations((), (), ()))
mne/io/fiff/raw.py:
self.set_annotations(Annotations((), (), ()))

you see that certain readers put empty Annotations on read. Some don't...

you also see that you have a number of "self.annotations.append"

I am still not sure myself what is the best decision but for sure we are
currently not consistent
between all Raw objects.

I am +1 on having empty annotations though when there is an annotations
block in the file on disk
for the argument you mention.

**Comment by larsoner** [comment]
> mne/io/fiff/raw.py:
> self.set_annotations(Annotations((), (), ()))

The only place I see this happening is when we are definitely adding new annotations to make note of data skips. Arguably the cleanest thing would actually be to check to see if these are the only type annotations present in the object, and if so, remove them before writing (as they will be recreated on read). But this seems like too much effort for what is a pretty rare occurrence in FIF data.

Thinking about it, I guess we have three options:

1. set `raw.annotations` based on whether on not it makes sense for the format (current behavior-ish). This is what is done now, and what I thought was being proposed. It means that for BrainVision, EEGLAB, presumably EGI and some others, it makes sense to *always* add annotations (even if len 0) since they all have native acquisition support for them. For FIFF there is no native (acq) support so it does not. This breaks consistency across formats (`None` vs `len(raw.annot) == 0`) but it makes round-trip I/O more consistent for FIF.

2. always set `raw.annot` in readers. This makes everything consistent, but breaks round-trip FIF I/O. To address that, we can either:

    1.  If `len(raw.annot) == 0` (and *maybe* someday also if the only thing it contains is FIF ACQ skips that we will reconstruct on read anyway), don't write the block. This might be the cleanest and simplest option actually, even though it might discard any `orig_time` information the user supplied.
    2. Add an argument `write_empty_annot=True` that people can set to `False` if they need to preserve round-trip I/O content as much as possible. I am wary of this option because if I do `mne.io.read_raw_fif(fname).save(fname_2)` the content in `fname` and `fname_2` should really be the same (up to ordering) if possible, and now it *almost never* will be unless I add `write_empty_annot=False`, which is a bit ugly. And setting `write_empty_annot=True` to the default is almost the same thing as (2i).

So I think my vote is for 2i.

**Comment by agramfort** [comment]
I also think that 2i is the simplest to maintain.
I am not worried about the orig_time info lost as it will be typically the
meas_date of the related raw
if any or None.

**Comment by mmagnuski** [comment]
I'm +1 on `2i` too. :)

File: mne/annotations.py [hunk 1]
```diff
@@ -188,11 +188,23 @@ def __len__(self):
...
+        Both annotations must have the same orig_time
+        """
+        if len(self) == 0:
+            self.orig_time = other.orig_time
+        if self.orig_time != other.orig_time:
+            raise ValueError("orig_time should be the same to "
+                             "add/concatenate 2 annotations "
+                             "(got %s != %s)" % (self.orig_time,
+                                                 other.orig_time))
+        # XXX this should not ignore orig_time
```
Comment (larsoner) [inline mne/annotations.py]: This comment seems outdated now

File: mne/channels/channels.py [hunk 2]
```diff
@@ -582,11 +582,10 @@ def anonymize(self):
         """
         anonymize_info(self.info)
         if hasattr(self, 'annotations'):
-            if self.annotations is None:
-                pass
-            else:
-                self.annotations.orig_time = None
-                self.annotations.onset -= self._first_time
+            # XXX : anonymize should rather subtract a random date
+            # rather than setting it to None
```
Comment (larsoner) [inline mne/channels/channels.py]: Why is that? This seems fine

Comment (agramfort) [reply mne/channels/channels.py]: that's what does mne_anonymize and what recommends the BIDS spec. One motivation is to change the date of a full study, eg to know in what is the order of the subjects and to know in which order the runs occurred. I wrote it down to not forget.

Comment (massich) [reply mne/channels/channels.py]: IMHO, set it to None is the most natural way and allows to check for anonymity. But I also understand the value of setting an arbitrary date. Then we would need to have a flag or something to indicate that the date was already altered for anonymity.

Either way, we should open an issue.

**Comment by agramfort** [comment]
thx @massich for the review

@larsoner good to go from your end?

File: mne/annotations.py (hunk 1 above)
Comment (massich) [reply mne/annotations.py]: ticket: #5555

**Comment by massich** [comment]
@agramfort, thanks for the code !

File: mne/viz/utils.py [hunk 3]
```diff
@@ -1952,7 +1945,7 @@ def _setup_annotation_colors(params):
     raw = params['raw']
     segment_colors = params.get('segment_colors', dict())
     # sort the segments by start time
-    if raw.annotations is not None:
+    if len(raw.annotations) > 0:
```
Comment (larsoner) [inline mne/viz/utils.py]: This conditional should no longer be necessary. `numpy.argsort` (and the `ndarray` method wrapper) should function correctly on empty arrays

File: mne/io/base.py [hunk 4]
```diff
@@ -2447,7 +2447,7 @@ def _start_writing_raw(name, info, sel=None, data_type=FIFF.FIFFT_FLOAT,
     #
     # Annotations
     #
-    if annotations is not None:  # allow saving empty annotations
+    if annotations is not None and len(annotations):  # don't save empty annot
```
Comment (larsoner) [inline mne/io/base.py]: Are there still conditions under which `annotations is None` here? It seems like there should not be, so this part of the conditional can be removed

**Comment by agramfort** [comment]
green @larsoner

File: mne/channels/channels.py (hunk 2 above)
Comment (jasmainak) [reply mne/channels/channels.py]: you should make the date before 1900 according to BIDS spec (if I remember correctly), so it's obvious it has been anonymized. The date is not completely arbitrary :)

**Pull request merged**